### PR TITLE
Add persistence cache

### PR DESCRIPTION
**Pull request by jantimon**
@numical what do you think?

### CODE REVIEW

**Comment by numical** [comment]
Ha! I knew you would be too quick for me.
I have had a chance to have a quick look at the file deltas but not enough time to download the code and take a closer look.  I'm afraid I'm still not confident enough on the compilation process to comment confidently without running the code. Will get back to you as soon as I can.
One immediate concern, which I hope is just my misunderstanding of the code.  If you do get a cache hit, are you not relying on the icon files themselves still existing?  What happens if they have cleared their build directory?

**Comment by jantimon** [comment]
This was actually intended - the cache file is also generated in the build folder. If you remove it it will generate the logo again. Do you think we have to verify that all images exist?

**Comment by numical** [comment]
Aha!  I think we have different use cases in mind.

OK - first of all to answer your question: Yes, we should verify referenced files exist.  My initial thought was 'why on earth would one or two go missing'.  But I think the more persuasive argument is that, if for whatever reason some did go missing, *not* checking could lead to some very hard to find bugs for DevOps.

On a more general point, I think we need to handle build processes where:
- the build directory cannot be 'polluted' with build artifacts such as the `.cache` file;
- the build directory is cleared prior to each run.
These are both features I have seen in many developments I have been involved in.

It would be easy enough to configure the location of the `.cache` file.  However what about storing cached copies of the image files? Sure we can do it but I think we would end up with caching concerns spread across 'index.js' and 'loader.js'.  Would it not be neater to have a file-based approach entirely in 'index.js' which simply bypasses the loading and compilation process when neither `options` nor `logo` file have changed?  (This also has the minor advantage of avoiding any JSON serialisation/deserialisation which I am always uncomfortable about). 

You may have noticed I have gone back to my suggestion from [here](https://github.com/jantimon/favicons-webpack-plugin/pull/1) ;-) 

If you see any mileage in this I will give it go.  If you are dead against, I will not bother.

**Comment by jantimon** [comment]
You are right the unexpected behaviour and untraceable devops issue is really a nogo.

Storing the cached file outside the dist directory is a cool idea - we would have to make sure that this doesn't conflict with similar compilations. E.g. if you would generate the same logo with different configurations.

What would be the advantage of copying the files instead of checking that they are still there?

I guess cleaning the dist directory should not be part of this plugin.

**Comment by numical** [comment]
**Copying vs checking for existing files** 
Actually, considering the unexpected and untraceable issues is what got me to the copying of the files, even if they are there.  The use cases around this area would normally be the edge cases!
e.g 1 : updating the logo might mean changing the contents but not the file names;
e.g 2: changing deployment options might mean simply adding/removing a subset of the image files;
e.g 3: another deployment step outside of webpack might also touch these files
Thus, instead of assuming that existing files are OK, I was planning to always replace in-situ files with ones I trust from my own cache store.

**Cache File conflicts**
I would not worry about this in a first implementation as long as the hash checks are based on the full `options` and `icon` file contents.  Simply store the last one.
If we feel that caching multiple parallel deployments might be useful we can then pretty easily convert the cache to a HashMap of cached data.

**Cleaning dist directory**
I do not mean we do it. I just mean we are aware others might.

**Speed**
My proposal of storing and copying a bunch of files is based on the bold assumption that the associated (async) IO would be less burdensome to the overall build process than running `favicon` every time; and not significantly slower than doing the necessary checks on existing files on a cache hit.  I could be horribly wrong.

**Comment by numical** [comment]
**tl;dr**
Too many words and not enough coding.  
Will give this a bash over the next couple of days. 
Will fork off 0.0.3 so you can then compare with what you have out into 0.0.4

**Comment by jantimon** [comment]
I guess we will need a json file in any case so we can add the html to the HtmlWebpackPlugin - so you might also start with 0.0.4 and just move the cache file ;)

**Comment by jantimon** [comment]
Oh just thought that if you use file.emit instead of copying it might be as fast as storing them in the dist directory.
However it wouldn't be easy to clear the cache.
